oslo/torch/nn/parallel/data_parallel/zero/sharded_optim/_utils.py

```diff
@@ -0,0 +1,284 @@
+        ):
+            return True
+        return False
+
+
+def release_param_grad(tensor_list: List[torch.Tensor]):
+    """
+    Release the gradients of a list of tensors.
+
+    Args:
+        tensor_list (List[torch.Tensor]): List of tensors to release gradients for.
+    """
+    for tensor in tensor_list:
+        tensor.grad = None
+
+
+def compute_norm(
+    gradients: Iterable[torch.Tensor],
+    params: Iterable[torch.Tensor],
+    dp_group: dist.ProcessGroup,
+    mp_group: dist.ProcessGroup,
+    norm_type: Union[int, float] = 2,
+) -> Iterable[torch.Tensor]:
+    """Clips gradient norm of an iterable of parameters.
+    This is adapted from torch.nn.utils.clip_grad.clip_grad_norm_ and
+    added functionality to handle model parallel parameters. Note that
+    the gradients are modified in place.
+
+    Args:
+        gradients (Iterable[Tensor]): an iterable of gradient tensors
+        params (Iterable[Tensor]): an iterable of parameters that the gradients are associated with
+        dp_group (torch.nn.parallel.ProcessGroup): data parallel process group
+        mp_group (torch.nn.parallel.ProcessGroup): model parallel process group
+         (float or int): max norm of the gradients
+        norm_type (Union[int, float]): type of the used p-norm. Can be ``2`` for L2 norm or ``inf`` for infinity norm. (default: 2)
+
+    Returns:
+        Iterable[torch.Tensor]: Total norm of the gradients (viewed as a single vector).
+    """
+
+    if mp_group is None:
+        mp_rank = 0
+    else:
+        mp_rank = dist.get_rank(mp_group)
+
+    norm_type = float(norm_type)
+    if norm_type == inf:
+        total_norm = max(g.data.abs().max() for g in gradients)
+        total_norm_cuda = torch.cuda.FloatTensor([float(total_norm)])
+        dist.all_reduce(
+            total_norm_cuda, op=torch.distributed.ReduceOp.MAX, group=dp_group
+        )
+
+        # Take max across all GPUs.
+        if mp_group is not None:
+            dist.all_reduce(tensor=total_norm_cuda, op=torch.distributed.ReduceOp.MAX)
+        total_norm = total_norm_cuda[0].item()
+    else:
+        total_norm = 0.0
+        # if dist.get_rank() == 0:
+        #    logger.info(f"Total Norm beginning {total_norm}")
+
+        for g, p in zip(gradients, params):
+            # Pipeline parallelism may replicate parameters. Avoid multi-counting.
+            if is_model_parallel_parameter(p) or mp_rank == 0:
+                param_norm = g.data.double().norm(2)
+                total_norm += param_norm.item() ** 2
+
+        # Sum across all model parallel GPUs.
+        total_norm_cuda = torch.cuda.FloatTensor([float(total_norm)])
+        torch.distributed.all_reduce(
+            total_norm_cuda, op=torch.distributed.ReduceOp.SUM, group=dp_group
+        )
+
+        if mp_group is not None:
+            dist.all_reduce(tensor=total_norm_cuda, op=torch.distributed.ReduceOp.SUM)
+
+        total_norm = total_norm_cuda[0].item() ** (1.0 / norm_type)
+
+    if (
```

Comment:
Can we use `has_inf_or_nan` here?

Reply:
I'm sorry, but it seems that this request is difficult to apply. `has_inf_or_nan` takes a tensor as input, converts it to float, and then checks the inf or nan of the elements. This syntax checks the inf and nan of the float `total_norm`.